File: AUTO3D_claude/js/geometry.js
/* AUTO3D - geometry.js : puntos de fuga, calibracion monocular,
   hipotesis de planos y reconstruccion 3D de un solo punto de vista. */
(function (A) {
  'use strict';
  var G = A.geom = {};

  /* Direccion (unitaria) desde el punto medio de un segmento hacia un vp
     homogeneo. Si vp esta en el infinito (w~0) la direccion es la propia
     del vp. */
  function dirToVP(seg, vp) {
    var mx = (seg.p[0] + seg.q[0]) / 2, my = (seg.p[1] + seg.q[1]) / 2;
    var dx, dy;
    if (Math.abs(vp[2]) < 1e-9) { dx = vp[0]; dy = vp[1]; }
    else { dx = vp[0] / vp[2] - mx; dy = vp[1] / vp[2] - my; }
    var n = Math.hypot(dx, dy) || 1;
    return [dx / n, dy / n];
  }
  function segDir(s) { var dx = s.q[0] - s.p[0], dy = s.q[1] - s.p[1], n = Math.hypot(dx, dy) || 1; return [dx / n, dy / n]; }

  function angErr(s, vp) {
    var a = segDir(s), b = dirToVP(s, vp);
    var c = Math.abs(a[0] * b[0] + a[1] * b[1]);
    return Math.acos(A.clamp(c, -1, 1));   // 0..pi/2
  }

  /* ---------- RANSAC de puntos de fuga ----------
     Devuelve hasta 3 vps homogeneos con sus segmentos asignados.
     Ponderado por longitud: un muro largo pesa mas que un ruido corto. */
  G.findVPs = function (segs, P, w, h) {
    var pool = segs.slice(0, 400).filter(function (s) { return s.len >= P.minSegLen; });
    var tol = A.rad(P.vpInlierDeg), res = [], used = new Array(pool.length).fill(false);
    var iters = P.vpIters;

    for (var vpi = 0; vpi < 3; vpi++) {
      var avail = [];
      for (var i = 0; i < pool.length; i++) if (!used[i]) avail.push(i);
      if (avail.length < 6) break;
      var best = null, bestScore = 0;
      for (var it = 0; it < iters; it++) {
        var a = avail[(Math.random() * avail.length) | 0];
        var b = avail[(Math.random() * avail.length) | 0];
        if (a === b) continue;
        var v = A.cross3(pool[a].line, pool[b].line);
        var nn = Math.hypot(v[0], v[1], v[2]); if (nn < 1e-9) continue;
        v = [v[0] / nn, v[1] / nn, v[2] / nn];
        var score = 0;
        for (var k = 0; k < avail.length; k++) {
          var s = pool[avail[k]];
          if (angErr(s, v) < tol) score += s.len;
        }
        if (score > bestScore) { bestScore = score; best = v; }
      }
      if (!best) break;
      // refinamiento: minimos cuadrados sobre los inliers (SVD 3x3 por potencia inversa simplificada)
      var inl = [];
      for (var k2 = 0; k2 < avail.length; k2++) {
        var s2 = pool[avail[k2]];
        if (angErr(s2, best) < tol) inl.push(avail[k2]);
      }
      if (inl.length < 3) break;
      best = G.refineVP(inl.map(function (ix) { return pool[ix]; }), best);
      var lenSum = 0;
      inl.forEach(function (ix) { used[ix] = true; lenSum += pool[ix].len; pool[ix].vp = vpi; });
      res.push({ v: best, idx: inl, weight: lenSum, n: inl.length });
    }
    res.sort(function (x, y) { return y.weight - x.weight; });
    // marcar de nuevo el indice de vp tras la ordenacion
    segs.forEach(function (s) { s.vp = -1; });
    res.forEach(function (r, i) { r.idx.forEach(function (ix) { pool[ix].vp = i; }); });
    return res;
  };

  /* Minimiza sum w_i (l_i . v)^2 con |v|=1  ->  autovector minimo de M=sum w l l^T.
     Iteracion inversa de potencia con 3 pasos de eliminacion gaussiana. */
  G.refineVP = function (segsIn, v0) {
    var M = [[0, 0, 0], [0, 0, 0], [0, 0, 0]];
    segsIn.forEach(function (s) {
      var l = s.line, n = Math.hypot(l[0], l[1]) || 1;
      var L = [l[0] / n, l[1] / n, l[2] / n], wgt = s.len;
      for (var i = 0; i < 3; i++) for (var j = 0; j < 3; j++) M[i][j] += wgt * L[i] * L[j];
    });
    // traza para regularizar
    var tr = (M[0][0] + M[1][1] + M[2][2]) / 3 || 1;
    var v = v0.slice();
    for (var it = 0; it < 12; it++) {
      var Ashift = [
        [M[0][0] - 1e-6 * tr, M[0][1], M[0][2]],
        [M[1][0], M[1][1] - 1e-6 * tr, M[1][2]],
        [M[2][0], M[2][1], M[2][2] - 1e-6 * tr]
      ];
      var x = solve3(Ashift, v);
      if (!x) break;
      var n2 = Math.hypot(x[0], x[1], x[2]); if (n2 < 1e-12) break;
      v = [x[0] / n2, x[1] / n2, x[2] / n2];
    }
    return v;
  };
  function solve3(Ain, bIn) {
    var M = [Ain[0].slice(), Ain[1].slice(), Ain[2].slice()], b = bIn.slice(), i, j, k;
    for (i = 0; i < 3; i++) {
      var piv = i;
      for (j = i + 1; j < 3; j++) if (Math.abs(M[j][i]) > Math.abs(M[piv][i])) piv = j;
      if (Math.abs(M[piv][i]) < 1e-14) return null;
      if (piv !== i) { var t = M[i]; M[i] = M[piv]; M[piv] = t; var tb = b[i]; b[i] = b[piv]; b[piv] = tb; }
      for (j = i + 1; j < 3; j++) {
        var f = M[j][i] / M[i][i];
        for (k = i; k < 3; k++) M[j][k] -= f * M[i][k];
        b[j] -= f * b[i];
      }
    }
    var x = [0, 0, 0];
    for (i = 2; i >= 0; i--) {
      var s = b[i];
      for (j = i + 1; j < 3; j++) s -= M[i][j] * x[j];
      x[i] = s / M[i][i];
    }
    return x;
  }

  /* ---------- calibracion a partir de 2-3 vps ortogonales ----------
     pp = centro de imagen.  (u-pp).(v-pp) = -f^2  para vps ortogonales.
     Devuelve {f, pp, dirs[], vertIdx} o null si no hay pareja valida. */
  G.calibrate = function (vps, w, h) {
    var pp = [w / 2, h / 2];
    var fin = [];
    vps.forEach(function (r, i) {
      if (Math.abs(r.v[2]) > 1e-7) fin.push({ i: i, p: [r.v[0] / r.v[2], r.v[1] / r.v[2]] });
    });
    var fs = [];
    for (var a = 0; a < fin.length; a++) for (var b = a + 1; b < fin.length; b++) {
      var ua = [fin[a].p[0] - pp[0], fin[a].p[1] - pp[1]];
      var ub = [fin[b].p[0] - pp[0], fin[b].p[1] - pp[1]];
      var d = ua[0] * ub[0] + ua[1] * ub[1];
      if (d < -1) fs.push(Math.sqrt(-d));
    }
    var f = fs.length ? A.median(fs) : null;
    if (!f || !isFinite(f)) return null;
    // direcciones 3D de cada vp: d = K^-1 v normalizado
    var dirs = vps.map(function (r) {
      var v = r.v, x, y, z;
      if (Math.abs(v[2]) > 1e-9) { x = v[0] / v[2] - pp[0]; y = v[1] / v[2] - pp[1]; z = f; }
      else { x = v[0]; y = v[1]; z = 0; }
      return A.normalize3([x, y, z]);
    });
    // vp vertical = aquel cuya direccion tiene mayor |componente y| en imagen
    var vertIdx = 0, bestV = -1;
    dirs.forEach(function (d, i) {
      var m = Math.abs(d[1]) / (Math.hypot(d[0], d[1]) || 1e-9);
      if (m > bestV) { bestV = m; vertIdx = i; }
    });
    // signo: "arriba" en camara = y negativa
    var up = dirs[vertIdx].slice();
    if (up[1] > 0) up = [-up[0], -up[1], -up[2]];
    return { f: f, pp: pp, dirs: dirs, vertIdx: vertIdx, up: up, nFinite: fin.length, fSamples: fs.length };
  };

  /* Horizonte = linea de fuga del plano horizontal. Se obtiene de la
     direccion vertical con l = K^-T . up, asi que basta con tener 2 vps
     (no hacen falta los 3).  Si hay dos vps horizontales se usa la recta
     que los une, que es mas estable. */
  G.horizon = function (vps, calib) {
    if (!calib) return null;
    var others = [];
    vps.forEach(function (r, i) { if (i !== calib.vertIdx) others.push(r.v); });
    if (others.length >= 2) {
      var l = A.cross3(others[0], others[1]);
      if (Math.hypot(l[0], l[1]) > 1e-9) return l;
    }
    var u = calib.up, f = calib.f, cx = calib.pp[0], cy = calib.pp[1];
    return [u[0] / f, u[1] / f, u[2] - (cx * u[0] + cy * u[1]) / f];
  };

  /* ---------- hipotesis de planos (cuadrilateros de 2 familias de vp) ----------
     Para cada par (i,j) de familias, se toman las K rectas mas largas de cada
     una, se intersecan y se puntua cada cuadrilatero por el soporte de borde
     real de sus 4 lados. Se quedan los mejores no solapados. */
  G.planeCandidates = function (det, vps, P) {
    var segs = det.segs, w = det.w, h = det.h, out = [];
    if (vps.length < 2) return out;
    var K = 9;
    function familyLines(fi) {
      var list = segs.filter(function (s) { return s.vp === fi; });
      list.sort(function (a, b) { return b.len - a.len; });
      return list.slice(0, K);
    }
    for (var i = 0; i < vps.length; i++) {
      for (var j = i + 1; j < vps.length; j++) {
        var A1 = familyLines(i), B1 = familyLines(j);
        if (A1.length < 2 || B1.length < 2) continue;
        for (var a1 = 0; a1 < A1.length; a1++) for (var a2 = a1 + 1; a2 < A1.length; a2++) {
          for (var b1 = 0; b1 < B1.length; b1++) for (var b2 = b1 + 1; b2 < B1.length; b2++) {
            var c = [
              A.interLines(A1[a1].line, B1[b1].line),
              A.interLines(A1[a1].line, B1[b2].line),
              A.interLines(A1[a2].line, B1[b2].line),
              A.interLines(A1[a2].line, B1[b1].line)
            ];
            if (c.indexOf(null) >= 0) continue;
            var ok = true;
            for (var t = 0; t < 4; t++) {
              if (c[t][0] < -w * 0.25 || c[t][0] > w * 1.25 || c[t][1] < -h * 0.25 || c[t][1] > h * 1.25) { ok = false; break; }
            }
            if (!ok) continue;
            var area = Math.abs(A.polyArea(c));
            if (area < w * h * 0.012) continue;
            if (area > w * h * 0.94) continue;
            var sup = 0;
            for (var e = 0; e < 4; e++) sup += A.vision.edgeSupport(det.edges, w, h, c[e], c[(e + 1) % 4], 2);
            sup /= 4;
            if (sup < P.planeSupport) continue;
            out.push({ pts: c, score: sup, area: area, fam: [i, j], id: A.uid('cand') });
          }
        }
      }
    }
    out.sort(function (x, y) { return (y.score * Math.sqrt(y.area)) - (x.score * Math.sqrt(x.area)); });
    // supresion de solapados por distancia de centroides + area
    var keep = [];
    for (var k = 0; k < out.length && keep.length < P.maxPlanes; k++) {
      var cand = out[k], cc = A.polyCentroid(cand.pts), dup = false;
      for (var m = 0; m < keep.length; m++) {
        var kc = A.polyCentroid(keep[m].pts);
        var rad = Math.sqrt(Math.max(cand.area, keep[m].area)) * 0.45;
        if (A.dist(cc, kc) < rad) { dup = true; break; }
      }
      if (!dup) keep.push(cand);
    }
    return keep;
  };

  /* ---------- reconstruccion monocular ----------
     Hipotesis de Manhattan + altura de camara conocida.
     Suelo: X.up = -camH  (camara en el origen, mirando -z? no: modelo pinhole
     estandar con z hacia delante).  Para un pixel p el rayo es r = K^-1 p.   */
  G.ray = function (calib, p) {
    return [p[0] - calib.pp[0], p[1] - calib.pp[1], calib.f];
  };
  G.groundPoint = function (calib, p, camH) {
    var r = G.ray(calib, p), d = A.dot3(r, calib.up);
    if (d > -1e-6) return null;                     // por encima del horizonte
    var t = -camH / d;
    if (t <= 0 || !isFinite(t)) return null;
    return A.mul3(r, t);
  };
  G.planeHit = function (calib, p, n, d) {          // n.X = d
    var r = G.ray(calib, p), den = A.dot3(r, n);
    if (Math.abs(den) < 1e-9) return null;
    var t = d / den;
    if (t <= 0 || !isFinite(t)) return null;
    return A.mul3(r, t);
  };

  /* Reconstruye un poligono segun su clase.
     - horizontal (suelo): todos los vertices sobre el plano del suelo.
     - vertical (fachada, hueco, medianera): se anclan los 2 vertices mas
       bajos al suelo y se define el plano vertical que los contiene.
     - libre (cubierta, otro): se intenta plano vertical; si falla, se marca
       como no reconstruible (se exporta solo en 2D).
     Devuelve {pts3:[[x,y,z]...], method, warn} o {error:'...'}          */
  G.reconstructPoly = function (pts, cls, calib, camH) {
    if (!calib) return { error: 'sin calibracion (hacen falta 2 puntos de fuga ortogonales)' };
    var info = A.classById(cls), axis = info.axis;
    var i;
    if (axis === 'horizontal') {
      var out = [];
      for (i = 0; i < pts.length; i++) {
        var X = G.groundPoint(calib, pts[i], camH);
        if (!X) return { error: 'vertice por encima del horizonte: no se puede apoyar en el suelo' };
        out.push(X);
      }
      return { pts3: out, method: 'suelo' };
    }
    // plano vertical: anclar por los dos vertices mas bajos de la imagen
    var order = pts.map(function (p, k) { return { p: p, k: k }; })
                   .sort(function (a, b) { return b.p[1] - a.p[1]; });
    var b1 = G.groundPoint(calib, order[0].p, camH);
    var b2 = null, used = 1;
    for (i = 1; i < order.length && !b2; i++) {
      var cand = G.groundPoint(calib, order[i].p, camH);
      if (cand && A.dist(order[0].p, order[i].p) > 8) { b2 = cand; used = i; }
    }
    if (!b1 || !b2) return { error: 'no hay dos vertices apoyables en el suelo (base oculta o sobre el horizonte)' };
    var dirH = A.normalize3(A.sub3(b2, b1));
    var n = A.normalize3(A.cross3(dirH, calib.up));   // normal del plano vertical
    var d = A.dot3(n, b1);
    var pts3 = [];
    for (i = 0; i < pts.length; i++) {
      var Xi = G.planeHit(calib, pts[i], n, d);
      if (!Xi) return { error: 'vertice paralelo al plano: geometria degenerada' };
      pts3.push(Xi);
    }
    var warn = (axis === 'libre') ? 'clase sin eje definido: se ha supuesto plano vertical' : null;
    return { pts3: pts3, method: 'vertical', normal: n, warn: warn };
  };

  /* cambio a coordenadas de mundo alineadas con Manhattan (X,Y arriba,Z) */
  G.toWorld = function (calib, X) {
    var up = calib.up, ax = null, az = null;
    calib.dirs.forEach(function (d, i) {
      if (i === calib.vertIdx) return;
      if (!ax) ax = d; else if (!az) az = d;
    });
    if (!ax) ax = A.normalize3([1, 0, 0]);
    if (!az) az = A.normalize3(A.cross3(up, ax));
    ax = A.normalize3(A.sub3(ax, A.mul3(up, A.dot3(ax, up))));   // ortogonalizar
    az = A.normalize3(A.cross3(up, ax));
    return [A.dot3(X, ax), A.dot3(X, up) + 0, A.dot3(X, az)];
  };
})(window.A3D);

File: AUTO3D_claude/python/vision_teach/geometry.py
"""Geometria de escena: aristas rectas, puntos de fuga, calibracion y 3D.

Complementa a `detection.py`. Alli una "arista" es un contorno de Canny y un
"plano" una mancha de color Lab; aqui una arista es un segmento recto asignado
a un punto de fuga, y un plano un cuadrilatero con normal en el espacio.

Supuestos: mundo de Manhattan (tres direcciones ortogonales dominantes), camara
pinhole sin distorsion y centro optico en el centro de la imagen. La escala
metrica viene de un unico dato externo: la altura de la camara sobre el suelo.

Sin dependencias fuera de numpy y OpenCV, que ya estan en requirements.txt.
"""
import numpy as np
import cv2

EPS = 1e-9


# --------------------------------------------------------------------------
# 1. De pixeles de borde a segmentos rectos
# --------------------------------------------------------------------------
def segments(frame, low, high, min_length=26., epsilon=1.8):
    """Devuelve (segmentos Nx4, mapa de bordes).

    Reutiliza el mismo Canny que `detection.detect`, de modo que los umbrales
    que aprende el Random Forest calibran tambien esta parte. Los contornos se
    parten en tramos rectos con approxPolyDP: a diferencia de HoughLinesP, cada
    tramo conserva sus extremos reales, que es lo que necesitan tanto el
    cuadrilatero de un plano como el seguimiento de un vertice.
    """
    gray = cv2.GaussianBlur(cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY), (5, 5), 0)
    edges = cv2.Canny(gray, int(low), int(high), L2gradient=True)
    contours, _ = cv2.findContours(edges, cv2.RETR_LIST, cv2.CHAIN_APPROX_NONE)
    out = []
    for contour in contours:
        if len(contour) < 3:
            continue
        polyline = cv2.approxPolyDP(contour, epsilon, False).reshape(-1, 2)
        for a, b in zip(polyline[:-1], polyline[1:]):
            if np.hypot(*(b - a)) >= min_length:
                out.append([a[0], a[1], b[0], b[1]])
    return np.asarray(out, dtype=np.float64).reshape(-1, 4), edges


def _parts(segs):
    """Precalcula recta homogenea, direccion unitaria, punto medio y longitud."""
    p, q = segs[:, :2], segs[:, 2:]
    ones = np.ones((len(segs), 1))
    lines = np.cross(np.hstack([p, ones]), np.hstack([q, ones]))
    delta = q - p
    length = np.linalg.norm(delta, axis=1)
    direction = delta / np.maximum(length, EPS)[:, None]
    return lines, direction, (p + q) / 2, length


def _cosines(vanishing, mids, directions):
    """|cos| del angulo entre cada segmento y la direccion hacia el punto de fuga."""
    if abs(vanishing[2]) > EPS:
        towards = vanishing[:2] / vanishing[2] - mids
    else:                                    # punto de fuga en el infinito
        towards = np.broadcast_to(vanishing[:2], mids.shape)
    towards = towards / np.maximum(np.linalg.norm(towards, axis=1), EPS)[:, None]
    return np.abs(np.sum(towards * directions, axis=1))


# --------------------------------------------------------------------------
# 2. Puntos de fuga
# --------------------------------------------------------------------------
def vanishing_points(segs, iterations=900, tolerance_deg=2.2, maximum=3, seed=0):
    """RANSAC en cascada, con el voto ponderado por longitud.

    Un muro de 300 px debe pesar mas que diez ruidos de 30 px: contar segmentos
    en vez de longitud deja que la textura fina decida la direccion dominante.
    """
    if len(segs) < 4:
        return []
    lines, directions, mids, lengths = _parts(segs)
    threshold = np.cos(np.radians(tolerance_deg))
    rng = np.random.default_rng(seed)          # reproducible entre ejecuciones
    available = np.ones(len(segs), bool)
    found = []

    for _ in range(maximum):
        index = np.flatnonzero(available)
        if len(index) < 6:
            break
        best, best_score = None, 0.
        for _ in range(iterations):
            a, b = rng.choice(index, 2, replace=False)
            candidate = np.cross(lines[a], lines[b])
            norm = np.linalg.norm(candidate)
            if norm < EPS:
                continue
            candidate = candidate / norm
            hits = _cosines(candidate, mids[index], directions[index]) > threshold
            score = lengths[index][hits].sum()
            if score > best_score:
                best_score, best = score, candidate
        if best is None:
            break
        hits = index[_cosines(best, mids[index], directions[index]) > threshold]
        if len(hits) < 3:
            break
        best = refine(lines[hits], lengths[hits])
        hits = index[_cosines(best, mids[index], directions[index]) > threshold]
        if len(hits) < 3:
            break
        available[hits] = False
        found.append({'point': best, 'inliers': hits, 'weight': float(lengths[hits].sum())})

    found.sort(key=lambda item: -item['weight'])
    return found


def refine(lines, weights):
    """Minimiza sum w (l.v)^2 con |v|=1: autovector menor de sum w l l^T."""
    normalized = lines / np.maximum(np.linalg.norm(lines[:, :2], axis=1), EPS)[:, None]
    scatter = (normalized * weights[:, None]).T @ normalized
    _, vectors = np.linalg.eigh(scatter)
    return vectors[:, 0]


def labels(segs, found):
    """Indice de punto de fuga por segmento, o -1 si no pertenece a ninguno."""
    out = np.full(len(segs), -1, int)
    for i, item in enumerate(found):
        out[item['inliers']] = i
    return out


# --------------------------------------------------------------------------
# 3. Calibracion y horizonte
# --------------------------------------------------------------------------
def calibrate(found, width, height):
    """Focal a partir de dos puntos de fuga ortogonales: (u-pp).(v-pp) = -f^2.

    Devuelve None si no hay ninguna pareja valida, que es lo que ocurre cuando
    se fotografia una fachada de frente: un solo punto de fuga no da escala.
    """
    principal = np.array([width / 2., height / 2.])
    finite = [(i, item['point'][:2] / item['point'][2])
              for i, item in enumerate(found) if abs(item['point'][2]) > 1e-7]
    focals = []
    for a in range(len(finite)):
        for b in range(a + 1, len(finite)):
            product = np.dot(finite[a][1] - principal, finite[b][1] - principal)
            if product < -1.:
                focals.append(np.sqrt(-product))
    if not focals:
        return None
    focal = float(np.median(focals))
    if not np.isfinite(focal) or focal <= 0:
        return None

    directions = []
    for item in found:
        point = item['point']
        if abs(point[2]) > EPS:
            vector = np.r_[point[:2] / point[2] - principal, focal]
        else:
            vector = np.r_[point[:2], 0.]
        directions.append(vector / max(np.linalg.norm(vector), EPS))
    directions = np.asarray(directions)

    # la vertical del mundo es la direccion mas vertical en la imagen
    vertical = int(np.argmax(np.abs(directions[:, 1]) /
                             np.maximum(np.linalg.norm(directions[:, :2], axis=1), EPS)))
    up = directions[vertical].copy()
    if up[1] > 0:                              # "arriba" es y negativa en imagen
        up = -up
    return {'focal': focal, 'principal': principal, 'directions': directions,
            'vertical': vertical, 'up': up, 'samples': len(focals)}


def horizon(found, calibration):
    """Linea de fuga del plano horizontal.

    Con tres puntos de fuga es la recta que une los dos horizontales. Con solo
    dos -el caso corriente- se deriva de la vertical: l = K^-T . up. La primera
    version exigia tres y dejaba el horizonte sin calcular casi siempre.
    """
    if calibration is None:
        return None
    others = [item['point'] for i, item in enumerate(found) if i != calibration['vertical']]
    if len(others) >= 2:
        line = np.cross(others[0], others[1])
        if np.hypot(line[0], line[1]) > EPS:
            return line
    up, focal = calibration['up'], calibration['focal']
    cx, cy = calibration['principal']
    return np.array([up[0] / focal, up[1] / focal,
                     up[2] - (cx * up[0] + cy * up[1]) / focal])


def horizon_y(line, width):
    """Altura del horizonte en el centro de la imagen, o None si es vertical."""
    if line is None or abs(line[1]) < EPS:
        return None
    return float(-(line[0] * width / 2. + line[2]) / line[1])


# --------------------------------------------------------------------------
# 4. Planos candidatos
# --------------------------------------------------------------------------
def edge_support(edges, a, b, tolerance=2):
    """Fraccion del segmento a-b que discurre sobre pixeles de borde reales.

    Es lo que separa un cuadrilatero apoyado en la fachada de otro que solo
    existe porque dos rectas se cruzan en el cielo.
    """
    height, width = edges.shape
    length = np.hypot(*(np.asarray(b) - np.asarray(a)))
    if length < 2:
        return 0.
    count = max(4, int(length / 2))
    t = np.linspace(0., 1., count + 1)
    xs = np.clip(np.round(a[0] + (b[0] - a[0]) * t).astype(int), 0, width - 1)
    ys = np.clip(np.round(a[1] + (b[1] - a[1]) * t).astype(int), 0, height - 1)
    hit = np.zeros(len(t), bool)
    for dy in range(-tolerance, tolerance + 1):
        for dx in range(-tolerance, tolerance + 1):
            hit |= edges[np.clip(ys + dy, 0, height - 1),
                         np.clip(xs + dx, 0, width - 1)] > 0
    return float(hit.mean())


def plane_candidates(segs, group, edges, support=.42, maximum=12, per_family=9):
    """Cuadrilateros formados por dos rectas de una familia y dos de otra.

    Se puntuan por soporte de borde, no por su area ni por lo bien que encajan
    entre si: un plano sin bordes que lo sostengan no esta en la fotografia.
    """
    height, width = edges.shape
    lines, _, _, lengths = _parts(segs)
    families = sorted({int(g) for g in group if g >= 0})
    found = []

    def longest(family):
        index = np.flatnonzero(group == family)
        return index[np.argsort(-lengths[index])][:per_family]

    for i in range(len(families)):
        for j in range(i + 1, len(families)):
            first, second = longest(families[i]), longest(families[j])
            for a1 in range(len(first)):
                for a2 in range(a1 + 1, len(first)):
                    for b1 in range(len(second)):
                        for b2 in range(b1 + 1, len(second)):
                            corners = _quad(lines, first[a1], first[a2], second[b1], second[b2])
                            if corners is None:
                                continue
                            if (corners[:, 0] < -width * .25).any() or (corners[:, 0] > width * 1.25).any():
                                continue
                            if (corners[:, 1] < -height * .25).any() or (corners[:, 1] > height * 1.25).any():
                                continue
                            area = abs(cv2.contourArea(corners.astype(np.float32)))
                            if not (width * height * .012 <= area <= width * height * .94):
                                continue
                            score = np.mean([edge_support(edges, corners[k], corners[(k + 1) % 4])
                                             for k in range(4)])
                            if score >= support:
                                found.append({'corners': corners, 'score': float(score),
                                              'area': float(area),
                                              'families': (families[i], families[j])})

    found.sort(key=lambda item: -item['score'] * np.sqrt(item['area']))
    kept = []
    for candidate in found:
        centre = candidate['corners'].mean(axis=0)
        radius = np.sqrt(max(candidate['area'], 1.)) * .45
        if all(np.hypot(*(centre - k['corners'].mean(axis=0))) >= radius for k in kept):
            kept.append(candidate)
            if len(kept) >= maximum:
                break
    return kept


def _quad(lines, a1, a2, b1, b2):
    corners = []
    for first, second in ((a1, b1), (a1, b2), (a2, b2), (a2, b1)):
        point = np.cross(lines[first], lines[second])
        if abs(point[2]) < 1e-12:
            return None
        corners.append(point[:2] / point[2])
    return np.asarray(corners)


# --------------------------------------------------------------------------
# 5. Reconstruccion metrica de una sola vista
# --------------------------------------------------------------------------
HORIZONTAL, VERTICAL, FREE = 'horizontal', 'vertical', 'libre'


def ray(calibration, point):
    return np.r_[point[0] - calibration['principal'][0],
                 point[1] - calibration['principal'][1], calibration['focal']]


def ground_point(calibration, point, camera_height):
    """Interseccion del rayo con el suelo, o None si mira por encima del horizonte."""
    direction = ray(calibration, point)
    denominator = float(np.dot(direction, calibration['up']))
    if denominator > -1e-6:
        return None
    distance = -camera_height / denominator
    return direction * distance if distance > 0 else None


def plane_point(calibration, point, normal, offset):
    direction = ray(calibration, point)
    denominator = float(np.dot(direction, normal))
    if abs(denominator) < EPS:
        return None
    distance = offset / denominator
    return direction * distance if distance > 0 else None


def reconstruct(points, axis, calibration, camera_height):
    """Lleva un poligono de la imagen al espacio metrico.

    La clase semantica decide la geometria: un suelo se apoya entero en el
    plano del terreno; una fachada se ancla por sus dos vertices mas bajos y
    el resto se proyecta sobre el plano vertical que pasa por ellos. Por eso
    etiquetar bien no es cosmetica.
    """
    if calibration is None:
        return {'error': 'sin calibracion: hacen falta dos puntos de fuga ortogonales'}
    points = np.asarray(points, dtype=np.float64)
    if len(points) < 3:
        return {'error': 'un plano necesita al menos tres vertices'}

    if axis == HORIZONTAL:
        space = []
        for point in points:
            found = ground_point(calibration, point, camera_height)
            if found is None:
                return {'error': 'vertice por encima del horizonte: no se apoya en el suelo'}
            space.append(found)
        return {'points': np.asarray(space), 'method': 'suelo'}

    order = np.argsort(-points[:, 1])            # de mas bajo a mas alto en imagen
    base = ground_point(calibration, points[order[0]], camera_height)
    other = None
    for index in order[1:]:
        candidate = ground_point(calibration, points[index], camera_height)
        if candidate is not None and np.hypot(*(points[index] - points[order[0]])) > 8:
            other = candidate
            break
    if base is None or other is None:
        return {'error': 'no hay dos vertices apoyables en el suelo (base oculta u horizonte)'}

    along = other - base
    along /= max(np.linalg.norm(along), EPS)
    normal = np.cross(along, calibration['up'])
    normal /= max(np.linalg.norm(normal), EPS)
    offset = float(np.dot(normal, base))
    space = []
    for point in points:
        found = plane_point(calibration, point, normal, offset)
        if found is None:
            return {'error': 'vertice paralelo al plano: geometria degenerada'}
        space.append(found)
    result = {'points': np.asarray(space), 'method': 'vertical', 'normal': normal}
    if axis == FREE:
        result['warning'] = 'clase sin eje definido: se ha supuesto plano vertical'
    return result


def world_axes(calibration):
    """Terna ortonormal alineada con el mundo de Manhattan (X, arriba, Z)."""
    up = calibration['up']
    horizontal = [d for i, d in enumerate(calibration['directions'])
                  if i != calibration['vertical']]
    first = horizontal[0] if horizontal else np.array([1., 0., 0.])
    first = first - up * np.dot(first, up)
    norm = np.linalg.norm(first)
    if norm < 1e-6:                              # degenerado: cualquier perpendicular sirve
        first = np.cross(up, [0., 0., 1.])
        norm = np.linalg.norm(first)
    first /= max(norm, EPS)
    return first, up, np.cross(up, first)


def to_world(calibration, points):
    first, up, third = world_axes(calibration)
    points = np.atleast_2d(points)
    return np.column_stack([points @ first, points @ up, points @ third])
